### PR TITLE
Initial commit of Pubsub GAPIC code.

### PR DESCRIPTION
**Pull request by chrisdunelm**
This is currently manually written code, but will ultimately be auto-generated.

### CODE REVIEW

File: src/Google.Pubsub.V1/project.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿{
+  "version": "0.1.0-*",
+  "summary": "GAPIC library for Google.Pubsub.V1",
```
Comment (jskeet) [inline src/Google.Pubsub.V1/project.json]: Let's just call it "Google Pubsub v1 client library"

File: src/Google.Pubsub.V1/project.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  "description": "GAPIC library for Google.Pubsub.V1, making common operations simpler in client code.",
```
Comment (jskeet) [inline src/Google.Pubsub.V1/project.json]: There's nothing really for it to be simpler than - I'd just duplicate the summary.

File: src/Google.Pubsub.V1/project.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  "authors": [ "Google Inc." ],
+  "owners": [ "google-apis-packages" ],
+  "tags": [ "Google" ],
+  "projectUrl": "https://github.com/GoogleCloudPlatform/gcloud-dotnet",
+  "licenseUrl": "https://developers.google.com/open-source/licenses/bsd",
+  "requireLicenseAcceptance": false,
+
```
Comment (jskeet) [inline src/Google.Pubsub.V1/project.json]: Nit - double line break (may be in storage too :)

File: src/Google.Pubsub.V1/SubscriberClient.cs [hunk 4]
```diff
@@ -0,0 +1,661 @@
...
+
+namespace Google.Pubsub.V1
+{
+    /// <summary>
+    /// Extension methods to assist with using <see cref="SubscriberClient"/>.
+    /// </summary>
+    public static partial class SubscriberExtensions
+    {
+        /// <summary>
+        /// Wrap a GRPC Publisher client for more convinient use.
```
Comment (jskeet) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Typo in convenient

File: src/Google.Pubsub.V1/SubscriberClient.cs [hunk 5]
```diff
@@ -0,0 +1,661 @@
...
+        /// </summary>
+        /// <param name="grpcClient">A GRPC client to wrap.</param>
+        /// <param name="settings">
+        /// An optional <see cref="SubscriberSettings"/> to configure this wrapper.
+        /// If null or not specified, then the default settings are used.
+        /// </param>
+        /// <returns>A <see cref="SubscriberSettings"/> that wraps the specified GRPC client.</returns>
+        public static SubscriberClient ToClient(
```
Comment (jskeet) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: We need to have another round of thinking about naming - definitely odd that SubscriberClient doesn't implement ISubscriberClient.

File: src/Google.Pubsub.V1/SubscriberClient.cs [hunk 6]
```diff
@@ -0,0 +1,661 @@
...
+            /// The default Subscriber scopes
+            /// </summary>
+            /// <remarks>
+            /// The default Subscriber scopes are:
+            /// <list type="bullet">
+            /// <item><description>"https://www.googleapis.com/auth/pubsub"</description></item>
+            /// <item><description>"https://www.googleapis.com/auth/cloud-platform"</description></item>
+            /// </list>
+            /// </remarks>
+            public static ReadOnlyCollection<string> Scopes { get; } = new ReadOnlyCollection<string>(new[] {
```
Comment (jskeet) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Ivan convinced me to use `IReadOnlyList<T>` in the public API for `PathTemplate`. We may want to do that here too, although I still don't like `IReadOnlyList` really...

File: src/Google.Pubsub.V1/SubscriberClient.cs [hunk 7]
```diff
@@ -0,0 +1,661 @@
...
+        /// <item><description>Port 443</description></item>
+        /// </list>
+        /// </remarks>
+        public static ServiceEndpointSettings GetDefaultServiceApiSettings() => new ServiceEndpointSettings
+        {
+            Host = ServiceDefaults.Host,
+            Port = ServiceDefaults.Port,
+        };
+
+        private static async Task<SubscriberClient> CreateFromDefaultCredentialsUnsafeAsync(
```
Comment (jskeet) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: When this is committed, I'll have a go at refactoring to put most of this into the GAX.

File: src/Google.Pubsub.V1/SubscriberClient.cs [hunk 8]
```diff
@@ -0,0 +1,661 @@
...
+        public static Task<SubscriberClient> CreateFromDefaultCredentialsAsync(
+            SubscriberSettings settings = null,
+            ServiceEndpointSettings serviceEndpointSettings = null,
+            IEnumerable<string> credentialScopes = null)
+        {
+            // Clone all settings, as Create can be async.
+            return CreateFromDefaultCredentialsUnsafeAsync(
+                settings?.Clone(),
+                serviceEndpointSettings?.Clone(),
+                credentialScopes == null ? null : new List<string>(credentialScopes));
```
Comment (jskeet) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Just `credentialScopes?.ToList()` would work here.

File: src/Google.Pubsub.V1/SubscriberClient.cs [hunk 9]
```diff
@@ -0,0 +1,661 @@
...
+            CancellationToken cancellationToken = default(CancellationToken),
+            Func<CallOptions, CallOptions> callOptionsOverride = null)
+        {
+            throw new NotImplementedException();
+        }
+
+        public virtual void Acknowledge(
+            string subscription,
+            IEnumerable<string> ackIds,
+            Func<CallOptions, CallOptions> callOptionsOverride = null)
```
Comment (jskeet) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Noted for further docs work.

File: src/Google.Pubsub.V1/SubscriberClient.cs [hunk 10]
```diff
@@ -0,0 +1,661 @@
...
+            }
+            else
+            {
+                IClock clock = settings.Clock ?? SystemClock.Instance;
+                TimeSpan timeout = settings.Timeout.Value;
+                this._calcDeadline = () => clock.GetCurrentDateTimeUtc() + timeout;
+            }
+        }
+
+        private readonly Func<DateTime?> _calcDeadline;
```
Comment (jskeet) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Code organization: fields at the top.

File: src/Google.Pubsub.V1/SubscriberClient.cs [hunk 11]
```diff
@@ -0,0 +1,661 @@
...
+                response => response.NextPageToken,
+                response => response.Subscriptions,
+                "" // An empty page-token
+            );
+
+        public SubscriberClientImpl(Subscriber.ISubscriberClient grpcClient, SubscriberSettings settings)
+        {
+            settings = settings ?? SubscriberSettings.GetDefault();
+            this.GrpcClient = grpcClient;
+            if (settings.Timeout == null)
```
Comment (jskeet) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Definitely something to put in GAX...

**Comment by jskeet** [comment]
Some nits, but otherwise happy. Not sure what to do about the Travis build...

File: src/Google.Pubsub.V1/project.json (hunk 1 above)
Comment (chrisdunelm) [inline src/Google.Pubsub.V1/project.json]: Done

File: src/Google.Pubsub.V1/project.json (hunk 2 above)
Comment (chrisdunelm) [inline src/Google.Pubsub.V1/project.json]: Done

File: src/Google.Pubsub.V1/project.json (hunk 3 above)
Comment (chrisdunelm) [inline src/Google.Pubsub.V1/project.json]: Done

File: src/Google.Pubsub.V1/SubscriberClient.cs (hunk 4 above)
Comment (chrisdunelm) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Done (and in PublisherClient.cs)

File: src/Google.Pubsub.V1/SubscriberClient.cs (hunk 5 above)
Comment (chrisdunelm) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Agreed.

File: src/Google.Pubsub.V1/SubscriberClient.cs (hunk 6 above)
Comment (chrisdunelm) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Done for the type. Concrete type is still a ReadOnlyCollection<string>.
(Slightly odd naming, that ReadOnlyCollection implements all the IList things)

File: src/Google.Pubsub.V1/SubscriberClient.cs (hunk 7 above)
Comment (chrisdunelm) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Agreed

File: src/Google.Pubsub.V1/SubscriberClient.cs (hunk 8 above)
Comment (chrisdunelm) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Done

File: src/Google.Pubsub.V1/SubscriberClient.cs (hunk 9 above)
Comment (chrisdunelm) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Agreed - I'll start this today

File: src/Google.Pubsub.V1/SubscriberClient.cs (hunk 10 above)
Comment (chrisdunelm) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Done

File: src/Google.Pubsub.V1/SubscriberClient.cs (hunk 11 above)
Comment (chrisdunelm) [inline src/Google.Pubsub.V1/SubscriberClient.cs]: Agreed

**Comment by jskeet** [comment]
LGTM if Travis goes green.

**Pull request merged**